src/nix/profile.cc

```diff
@@ -149,6 +153,34 @@ struct ProfileManifest
                         e["outputs"].get<ExtendedOutputsSpec>()
                     };
                 }
+
+                std::string nameCandidate;
+                if (e.contains("name")) {
+                    nameCandidate = e["name"];
+                }
+                else {
+                    /* Heuristically determine a decent name for the element. */
+                    if (element.source) {
+                        nameCandidate = std::regex_replace(static_cast<std::string>(e["attrPath"]), matchPackagesPrefix, "");
+                        if (nameCandidate == "default") {
+                            auto originalRef = element.source->originalRef;
+                            nameCandidate = originalRef.input.getName();
+                            if (nameCandidate == "source") {
+                                nameCandidate = std::regex_replace(originalRef.to_string(), matchFlakeUrlPrefix , "");
+                            }
+                        }
+                    }
+                    else {
+                        nameCandidate = element.identifier();
+                    }
```

Comment:
I made some changes to have getNameFromURL in a separate file that can be unit tested. You can see the changes here: https://github.com/bobvanderlinden/nix/compare/0c4c043369445ae77c0627d39a04d9b699d11908..18d3aefc09342740e34300452b1fc88220bfd549

Reply:
Awesome, that is indeed much easier to test! I'll incorporate this, but it does need some improvements.

Reply:
@bobvanderlinden I cherry-picked your commits, rebased everything on master and added more test cases. I didn't get around to making them pass yet, I just went through [these docs](https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-flake#url-like-syntax) and tried to find edge-cases that need to be handled.

If you have time, feel free to take a stab at it, otherwise I'll continue when I'm free again.

Reply:
I have worked in this for a bit and I remember making a change to url parsing to support #. I'll have a look tomorrow.

Reply:
@iFreilicht by branch includes the fixes: https://github.com/bobvanderlinden/nix/commits/profile-names-instead-of-index

https://github.com/bobvanderlinden/nix/commit/ac5874cf12ab827e569e97fbbc0bf6cd9586b90d fixes the URL parsing so that it accepts `^` (I had mistaken `#` to be the culprit, but it was `^`).

https://github.com/bobvanderlinden/nix/commit/e4e828d00d7aca93c8c34890d347795fdb743dc0 adds support for `Index` again (showing it in the `nix profile list` and referring to it using the index as argument).

https://github.com/bobvanderlinden/nix/commit/9575cddf3fe0f95892b6563c9a1d9eeae46dd600 improves on the previous regex implementation by instantiating the regex patterns once. Slight performance improvement probably.

Reply:
I looked at your commits now, and it doesn't seem like they help with the `getNameFromUrl` function. I'm also not sure the change regarding `#` and `^` is valid, I don't think you can install only a specific output of a flake. Or am I misunderstanding what you're trying to do here?

I already put `Index` in myself, so I skipped that one.

The last one is useful though,  I'll build upon that to support all the other schemes.

Reply:
Ok I understand now what the first one was about. I incorporated that one as well and added additional tests for these kinds of URLs as well.